File: Projects/CSharp/CustomPropertiesSample/SchemaCollections/TableConfigurationSerializer.cs
//------------------------------------------------------------------------------
//
// Copyright (c) 2002-2009 CodeSmith Tools, LLC.  All rights reserved.
// 
// The terms of use for this software are contained in the file
// named sourcelicense.txt, which can be found in the root of this distribution.
// By using this software in any fashion, you are agreeing to be bound by the
// terms of this license.
// 
// You must not remove this notice, or any other, from this software.
//
//------------------------------------------------------------------------------

using System;
using System.Xml;
using System.Xml.XPath;
using CodeSmith.Engine;
using CodeSmith.Engine.Schema;
using SchemaExplorer;
using SchemaExplorer.Serialization;

namespace CodeSmith.Samples
{
    public class TableConfigurationSerializer : IPropertySerializer
    {

        public object SaveProperty(PropertySerializerContext context, object propertyValue)
        {
            return propertyValue;
        }

        public object LoadProperty(PropertySerializerContext context, object propertyValue)
        {
            return propertyValue;
        }

        public void WritePropertyXml(PropertySerializerContext context, XmlWriter writer, object propertyValue)
        {
            TableConfiguration configuration = propertyValue as TableConfiguration;

            if (configuration != null)
            {
                DatabaseSchema database = null;
                if (configuration.SourceTable != null)
                {
                    database = configuration.SourceTable.Database;
                }
                else if (configuration.SourceView != null)
                {
                    database = configuration.SourceView.Database;
                }

                if (database != null)
                {
                    writer.WriteStartElement("DatabaseSchema");
                    new DatabaseSchemaSerializer().WritePropertyXml(context, writer, database);
                    writer.WriteEndElement();
                }

                if (configuration.SourceTable != null)
                {
                    writer.WriteStartElement("SourceTable");
                    writer.WriteElementString("Owner", configuration.SourceTable.Owner);
                    writer.WriteElementString("Name", configuration.SourceTable.Name);
                    writer.WriteEndElement();
                }

                if (configuration.SourceView != null)
                {
                    writer.WriteStartElement("SourceView");
                    writer.WriteElementString("Owner", configuration.SourceView.Owner);
                    writer.WriteElementString("Name", configuration.SourceView.Name);
                    writer.WriteEndElement();
                }
            }
        }

        public object ReadPropertyXml(PropertySerializerContext context, XmlNode propertyValue)
        {
            if (context.PropertyInfo.PropertyType != typeof(TableConfiguration))
            {
                return null;
            }
            return ReadPropertyXmlInner(context, propertyValue);
        }

        internal TableConfiguration ReadPropertyXmlInner(PropertySerializerContext context, XmlNode propertyValue)
        {
            XPathNavigator navigator = propertyValue.CreateNavigator();
            XmlNamespaceManager manager = new XmlNamespaceManager(navigator.NameTable);
            XPathNodeIterator iterator;

            // Add the CodeSmith namespace
            manager.AddNamespace("cs", CodeSmithProject.DefaultNamespace);

            // Get the DatabaseSchema
            DatabaseSchema databaseSchema = null;
            iterator = (XPathNodeIterator)navigator.Evaluate("cs:DatabaseSchema", manager);
            if (iterator.MoveNext())
            {
                XmlNode oNode = ((IHasXmlNode)iterator.Current).GetNode();
                databaseSchema = (DatabaseSchema)new DatabaseSchemaSerializer().ReadPropertyXml(context, oNode);
            }

            // Get the SourceTable
            TableSchema sourceTable = null;
            if (databaseSchema != null)
            {
                string strOwner = navigator.Evaluate("string(cs:SourceTable/cs:Owner/text())", manager) as string;
                string strName = navigator.Evaluate("string(cs:SourceTable/cs:Name/text())", manager) as string;

                if (!string.IsNullOrEmpty(strName))
                {
                    sourceTable = !string.IsNullOrEmpty(strOwner) ? databaseSchema.Tables[strOwner, strName] : databaseSchema.Tables[strName];
                }
            }

            // Get the SourceView
            ViewSchema sourceView = null;
            if (databaseSchema != null)
            {
                string strOwner = navigator.Evaluate("string(cs:SourceView/cs:Owner/text())", manager) as string;
                string strName = navigator.Evaluate("string(cs:SourceView/cs:Name/text())", manager) as string;

                if (!string.IsNullOrEmpty(strName))
                {
                    sourceView = !string.IsNullOrEmpty(strOwner) ? databaseSchema.Views[strOwner, strName] : databaseSchema.Views[strName];
                }
            }

            // Create and return
            return new TableConfiguration
                       {
                SourceTable = sourceTable,
                SourceView = sourceView
            };
        }

        public object ParseDefaultValue(PropertySerializerContext context, string defaultValue)
        {
            return null;
        }
    }
}

File: Projects/VisualBasic/CustomPropertiesSample/SchemaCollections/TableConfigurationSerializer.vb
'------------------------------------------------------------------------------
'
' Copyright (c) 2002-2011 CodeSmith Tools, LLC.  All rights reserved.
' 
' The terms of use for this software are contained in the file
' named sourcelicense.txt, which can be found in the root of this distribution.
' By using this software in any fashion, you are agreeing to be bound by the
' terms of this license.
' 
' You must not remove this notice, or any other, from this software.
'
'------------------------------------------------------------------------------

Imports System
Imports System.Xml
Imports System.Xml.XPath
Imports CodeSmith.Engine
Imports CodeSmith.Engine.Schema
Imports SchemaExplorer
Imports SchemaExplorer.Serialization

Namespace CodeSmith.Samples

    Public Class TableConfigurationSerializer
        Implements IPropertySerializer

        Public Function SaveProperty(ByVal context As PropertySerializerContext, ByVal propertyValue As Object) As Object Implements IPropertySerializer.SaveProperty
            Return propertyValue
        End Function

        Public Function LoadProperty(ByVal context As PropertySerializerContext, ByVal propertyValue As Object) As Object Implements IPropertySerializer.LoadProperty
            Return propertyValue
        End Function

        Public Sub WritePropertyXml(ByVal context As PropertySerializerContext, ByVal writer As XmlWriter, ByVal propertyValue As Object) Implements IPropertySerializer.WritePropertyXml
            Dim configuration As TableConfiguration = TryCast(propertyValue, TableConfiguration)

            If Not configuration Is Nothing Then
                Dim database As DatabaseSchema = Nothing
                If Not configuration.SourceTable Is Nothing Then
                    database = configuration.SourceTable.Database
                ElseIf Not configuration.SourceView Is Nothing Then
                    database = configuration.SourceView.Database
                End If

                If Not database Is Nothing Then
                    writer.WriteStartElement("DatabaseSchema")
                    Dim dss As New DatabaseSchemaSerializer()
                    dss.WritePropertyXml(context, writer, database)
                    writer.WriteEndElement()
                End If

                If Not configuration.SourceTable Is Nothing Then
                    writer.WriteStartElement("SourceTable")
                    writer.WriteElementString("Owner", configuration.SourceTable.Owner)
                    writer.WriteElementString("Name", configuration.SourceTable.Name)
                    writer.WriteEndElement()
                End If

                If Not configuration.SourceView Is Nothing Then
                    writer.WriteStartElement("SourceView")
                    writer.WriteElementString("Owner", configuration.SourceView.Owner)
                    writer.WriteElementString("Name", configuration.SourceView.Name)
                    writer.WriteEndElement()
                End If
            End If
        End Sub

        Public Function ReadPropertyXml(ByVal context As PropertySerializerContext, ByVal propertyValue As XmlNode) As Object Implements IPropertySerializer.ReadPropertyXml
            If Not context.PropertyInfo.PropertyType.ToString().Equals(GetType(TableConfiguration).ToString(), StringComparison.InvariantCultureIgnoreCase) Then
                Return Nothing
            End If
            Return ReadPropertyXmlInner(context, propertyValue)
        End Function

        Friend Function ReadPropertyXmlInner(ByVal context As PropertySerializerContext, ByVal propertyValue As XmlNode) As TableConfiguration
            Dim navigator As XPathNavigator = propertyValue.CreateNavigator()
            Dim manager As New XmlNamespaceManager(navigator.NameTable)
            Dim iterator As XPathNodeIterator

            ' Add the CodeSmith namespace
            manager.AddNamespace("cs", CodeSmithProject.DefaultNamespace)

            ' Get the DatabaseSchema
            Dim databaseSchema As DatabaseSchema
            iterator = DirectCast(navigator.Evaluate("cs:DatabaseSchema", manager), XPathNodeIterator)
            If iterator.MoveNext() Then
                Dim oNode As XmlNode = (DirectCast(iterator.Current, IHasXmlNode)).GetNode()
                databaseSchema = DirectCast(New DatabaseSchemaSerializer().ReadPropertyXml(context, oNode), DatabaseSchema)
            End If

            ' Get the SourceTable
            Dim sourceTable As TableSchema = Nothing
            If Not databaseSchema Is Nothing Then
                Dim strOwner As String = TryCast(navigator.Evaluate("string(cs:SourceTable/cs:Owner/text())", manager), String)
                Dim strName As String = TryCast(navigator.Evaluate("string(cs:SourceTable/cs:Name/text())", manager), String)

                If Not String.IsNullOrEmpty(strName) Then
                    sourceTable = If(Not String.IsNullOrEmpty(strOwner), databaseSchema.Tables(strOwner, strName), databaseSchema.Tables(strName))
                End If
            End If

            ' Get the SourceView
            Dim sourceView As ViewSchema = Nothing
            If Not databaseSchema Is Nothing Then
                Dim strOwner As String = TryCast(navigator.Evaluate("string(cs:SourceView/cs:Owner/text())", manager), String)
                Dim strName As String = TryCast(navigator.Evaluate("string(cs:SourceView/cs:Name/text())", manager), String)

                If Not String.IsNullOrEmpty(strName) Then
                    sourceView = If(Not String.IsNullOrEmpty(strOwner), databaseSchema.Views(strOwner, strName), databaseSchema.Views(strName))
                End If
            End If

            ' Create and return
            Return New TableConfiguration()
        End Function

        Public Function ParseDefaultValue(ByVal context As PropertySerializerContext, ByVal defaultValue As String) As Object Implements IPropertySerializer.ParseDefaultValue
            Return Nothing
        End Function
    End Class

End Namespace